dashboard/app/controllers/activities_controller.rb

```diff
@@ -56,6 +56,14 @@ def milestone
       end
     end
 
+    if current_user && @script_level && @script_level.stage.lockable?
+      user_level = UserLevel.find_by(user_id: current_user.id, level_id: @script_level.level.id, script_id: @script_level.script.id)
+      # we have a lockable stage, and user_level is locked. disallow milestone requests
+      if user_level.nil? || user_level.locked?(@script_level.stage)
+        return head 403
+      end
+    end
```

Comment:
I think we expect `LevelGroup` to be only used in non-cacheable scripts, but I don't know that this is enforced in code.